libs/core/src/bloom_filter.cpp

```diff
@@ -0,0 +1,375 @@
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/bitvector.hpp"
+#include "core/bloom_filter.hpp"
+#include "core/byte_array/const_byte_array.hpp"
+#include "crypto/fnv.hpp"
+
+#include <openssl/evp.h>
+
+#include <cassert>
+#include <cstddef>
+#include <functional>
+#include <vector>
+
+// Aim for 1 false positive per this many positive queries
+constexpr std::size_t const INVERSE_TARGET_FALSE_POSITIVE_RATE = 100000u;
+
+// No point in evaluating the filter's quality until this many positive queries
+// have been executed
+constexpr std::size_t const MEANINGFUL_STATS_THRESHOLD = 5u * INVERSE_TARGET_FALSE_POSITIVE_RATE;
+
+constexpr std::size_t const INITIAL_SIZE_IN_BITS = 8 * 10 * 1024 * 1024;
+
+namespace fetch {
+namespace internal {
+
+HashSourceFactory::HashSourceFactory(HashSourceFactory::Functions hash_functions)
+  : hash_functions_(std::move(hash_functions))
+{}
+
+HashSourceFactory::~HashSourceFactory() = default;
+
+HashSource HashSourceFactory::operator()(Bytes const &element) const
+{
+  return HashSource(this->hash_functions_, element);
+}
+
+HashSource::HashSource(HashSourceFactory::Functions const &hash_functions,
+                       HashSourceFactory::Bytes const &    input)
+  : data_{}
+{
+  // TODO(WK): lazily evaluate hashes
+  for (auto const &fn : hash_functions)
+  {
+    auto const hashes = fn(input);
+    data_.insert(data_.cend(), hashes.cbegin(), hashes.cend());
+  }
+}
+
+HashSource::~HashSource()                      = default;
+HashSource::HashSource(HashSource &&) noexcept = default;
+HashSource &HashSource::operator=(HashSource &&) noexcept = default;
+
+HashSource::HashSourceIterator HashSource::begin() const
+{
+  return HashSource::HashSourceIterator{this, 0u};
+}
+
+HashSource::HashSourceIterator HashSource::end() const
+{
+  return HashSource::HashSourceIterator{this, data_.size()};
+}
+
+HashSource::HashSourceIterator HashSource::cbegin() const
+{
+  return HashSource::HashSourceIterator{this, 0u};
+}
+
+HashSource::HashSourceIterator HashSource::cend() const
+{
+  return HashSource::HashSourceIterator{this, data_.size()};
+}
+
+std::size_t HashSource::getHash(std::size_t index) const
+{
+  return data_[index];
+}
+
+HashSource::HashSourceIterator::~HashSourceIterator() = default;
```

Comment:
All these `= default`s should be the headers

Reply:
Why? I think the opposite is true. This is implementation just like an explicit function body.

Reply:
Regarding our conversation: this is now done.